src/platform_impl/linux/wayland/event_loop/mod.rs

```diff
@@ -209,6 +262,37 @@ impl EventLoop {
 
             PumpStatus::Exit(code)
         } else {
+            // NOTE: spawn a wake-up thread, thus if we have code reading the wayland connection
+            // in parallel to winit, we ensure that the loop itself is marked as having events.
+            if timeout.is_some() && self.pump_event_notifier.is_none() {
+                let (event_loop_awakener, event_loop_awakener_source) =
+                    calloop::ping::make_ping().map_err(|err| os_error!(err)).unwrap();
+
+                self.pump_event_notifier = spawn_wakeup_thread(
+                    self.active_event_loop.handle.connection.clone(),
+                    event_loop_awakener,
+                );
+
+                if self.pump_event_notifier.is_some() {
+                    let wayland_dispatcher = self.wayland_dispatcher.clone();
+                    self.event_loop
+                        .handle()
+                        .insert_source(
+                            event_loop_awakener_source,
+                            move |_, _, winit_state: &mut WinitState| {
+                                let mut source = wayland_dispatcher.as_source_mut();
+                                let _ = winit_state.dispatch_pending(source.queue());
```

Comment:
This seems unnecessary.

Reply:
it's necessary. The dispatch isn't done otherwise IIRC, but I can test one more time though.

Reply:
Doesn't `pump_app_events` always dispatch events?

Reply:
Yeah, it's required. I think it's a bug though, should be fixed on `calloop` end.